chore(python): add type hints and minor changes

ProbstDJakob opened the pull request:
## Important: this PR only exists to split up the PR #208

- add python 3.6+ type hints
- replace class variables with instance variables (LDAPBackend, _LDAPUser, LDAPSettings)
- LDAPBackend and LDAPSettings have backwards compatibility for class variables
- renaming LDAPSettings._name(str) to LDAPSettings._prepend_prefix(str)
- replacing AUTH_LDAP_xyz in strings with LDAPSettings._prepend_prefix(xyz)
- _LDAPUser._normalize_mirror_settings() no longer edits LDAPSettings and instead returns the normalized Sets
- replace defaults dict in LDAPSettings with actual variables, allowing type hints, checking if variable exists (e.g. IDE), and preventing supplied defaults dict (LDAPSettings.__init__) to inject undesired variables
- add AbstractLDAPSearch as abstract base class
- add check for results is None in LDAPSearch.execute
- remove LDAPSearch._begin(...) and LDAPSearch._results(...)
- add conversion to string in LDAPSearch._escape_filterargs
- rename AbstractLDAPSearch._begin(...) to AbstractLDAPSearch._search(...) and AbstractLDAPSearch._results(...) to AbstractLDAPSearch._result(...)
- add AbstractLDAPSearch._abandon(...)
- LDAPSearch._search(...) saves the msgid inside the object instead of returning it
- make LDAPGroupType abstract
- MemberDNGroupType.is_member(...) now returns bool instead of Union[bool, int]
- remove kwargs from LDAPGroupQuery.__init__
- update test mocks corresponding to the AbstractLDAPSearch changes

francoisfreitag commented on dev-requirements.txt:
```diff
@@ -2,5 +2,5 @@ black
 flake8
 mock >= 2.0.0
-python-ldap >= 3.0
+python-ldap >= 3.1
```
Shouldn’t be needed for this PR.
```suggestion
python-ldap >= 3.0
```

ProbstDJakob replied:
Yes, and I added `isort`, `mypy`, and `tox` to the dev dependencies as well, but after #228 has been merged, this would not be necessary any more.

francoisfreitag commented on django_auth_ldap/__init__.py:
```diff
@@ -1,2 +1,2 @@
-VERSION = (2, 2, 0)
+VERSION = (2, 2, 1)
```
Shouldn’t be needed for this PR.
```suggestion
VERSION = (2, 2, 0)
```

ProbstDJakob replied:
This has been removed.

francoisfreitag commented on django_auth_ldap/backend.py:
```diff
@@ -45,94 +45,146 @@
 )
 
-logger = _LDAPConfig.get_logger()
+try:
+    from django.contrib.auth.backends import BaseBackend
+except ImportError:
+    BaseBackend = object
```
Why guard against import errors? We can import from django above, `BaseBackend` should always be importable.

ProbstDJakob replied:
Since the `BaseBackend` was first introduced in [Django 3.0](https://docs.djangoproject.com/en/dev/releases/3.0/#django-contrib-auth) and `django-auth-ldap` supports Django 2.2, this would result in an `ImportError` otherwise.

francoisfreitag replied:
I see, thanks for explaining. This can definitely use a comment, so that the import error catching is removed when support for Django 2.2 is removed.

ProbstDJakob replied:
In form of a TODO or just a simple comment?

francoisfreitag replied:
Both work, just something we can find when grepping `2.2`.

For example:
```python
# Stop catching ImportError when dropping support for Django 2.2.
```

francoisfreitag commented on django_auth_ldap/backend.py:
```diff
@@ -45,94 +45,146 @@
+    ) -> None:
+        # Check hasattr for support of previous class variables.
+
+        if not hasattr(self, "_settings"):
+            self.supports_anonymous_user: bool = supports_anonymous_user
+        if not hasattr(self, "_settings"):
+            self.supports_object_permissions: bool = supports_object_permissions
+        if not hasattr(self, "_settings"):
+            self.supports_inactive_user: bool = supports_inactive_user
+
+        if not hasattr(self, "_settings"):
+            self._settings: Optional[LDAPSettings] = settings
```
These `if` can be combined, no need to repeat the check.

ProbstDJakob replied:
Thanks for the hint. No they should not be combined neither should they all check `hasattr(self, "_settings")`, instead they should check for their corresponding attribute.

francoisfreitag reviewed (commented):
Could you rebase on the latest master and configure a mypy tox build to validate the type annotations? Just running mypy locally shows 59 errors.

ProbstDJakob commented:
> Could you rebase on the latest master and configure a mypy tox build to validate the type annotations? Just running mypy locally shows 59 errors.

I rebased the onto `master` and configured `mypy` and fixed the issues. This has been done in a later commit of #208, but already included `ldap3`, thus it has not been included in this PR at first.

francoisfreitag reviewed (commented):
Can you please split the PR further, so that this PR is only focused on adding type hints. There are good changes in the suggested “minor changes”, but they should be proposed separately. The history (and reviews) are much easier to follow when changes are small, accompanied by a rationale explaining why the changes bring an improvement. It also facilitates bisecting later issues.

ProbstDJakob commented on tox.ini:
```diff
@@ -32,6 +36,12 @@ deps = isort>=5.0.1
 skip_install = true
 
+[testenv:mypy]
+basepython = python3
```
This slipped throw while merging the two commits and rebasing. I will adjust this later.

francoisfreitag replied:
No problem, thanks for updating the branch and type hints so quickly, it’s definitely easier to review when I can experiment with the code :slightly_smiling_face:.

ProbstDJakob commented:
I've created a new PR (#231) to spilt this PR further up. This PR still contains some minor changes (AbstractLDAPSearch and LDAPSettings) in order to achive "good" type hints.
